### ISSUE TITLE
Mapping corruption during concurrent access in MAP_SHARED mode

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/libcx#68]
It looks like there is some race when several processses access the same file mapping at the same time in MAP_SHARED mode. See https://github.com/bitwiseworks/qtbase-os2/issues/72 for the real life example.

**Comment by dmik** [comment]
It's really hard to track because it's not really clear where it goes out of sync. What I see is that sometimes several moc processes try to map the same source file at (nearly) the same time so their exception handlers overlap. E.g.:

1. P1 maps some source file in R/O mode with MAP_SHARED. It's the first one, so a new shared region is allocated and returned to P1.
2. P1 then sequentially reads all bytes from that region at once into its own buffer.
3. Reading from each subsequent page causes the LIBCx mmap excetpion handler to kick in.
4. The exception handler reads the respective page from the file into the shared region's page.
5. Reading big files (several dozen pages) this way takes some time.
6. Somewhere in the middle of this reading sequence, another process, P2, kicks in with its own mapping of the same file. Since it's also MAP_SHARED, the very same shared region gets returned to P2.
7. Pages that have been already loadded by P1, are read by P2 instantly w/o triggering the mmap exception handler.
8. At some point P1 and P2 happen to access the same page which is not yet mapped. P2 wins this race, grabs the global LIBCx mutex, commits the page, reads the file contents into it and releases the lock. Then continues execution.
9. P1, while still being in its own mmap exception handler, finaly gets the lock, disovers that the page has already been committed (and read) and just continues execution.
10. Steps 8-9 repeat for the rest of the pages with P2 always winning the race.

In theory all should work. However, as practice shows, P1 ends up with incorrect file contents. In case of moc (https://github.com/bitwiseworks/qtbase-os2/issues/72), this breaks namespace termination detection which results into the wrong namespace (`std::`) to be appended to generated classes.

One of my wild guesses is that some internal cache isn't yet flushed when P1 gets access to data read by P2 so it sees either zeroes or some garbage. I will try to dig/test some more.

**Comment by dmik** [comment]
I've tried to force file reading in P1 anyway, even if it sees that P2 already did it, but this doesn't help. So I'm not sure it's caching any more.

**Comment by dmik** [comment]
So far, it's not clear where the race takes place. Indeed, mapping the same file happens quite often across moc processes. It even happens that the exception handler for the same uncommitted page in shared memory (whcih is about to get committed and read from disk in this handler) gets triggered at the same time in more than one processes. I.e. both processes see the exception but only one wins this race and does the commit + read job. The other one waits on the global shared mutex and, once it is released by the first process, notices that the page is committed and simply retries.

I've even checked that both processes see exactly the same bytes, at least, in both cases they match file contents. Given that there is no other potentially conflicting place (everything sensitive is also guarded by the globel mutex), I'm not sure what goes wrong and where. Judgihg from the logs, all parts work as it should.

**Comment by dmik** [comment]
I've just tried to enable only one core under the ArcaOS SMP kernel (14.201) with:
```
PSD=ACPI.PSD /MAXCPU=1
```
and this makes the problem completely go away. I tried to run the test in a cycle of 20 iterations and not a single faliure. With multi cores enabled, It fails on the 3-5th run. This is a good proof that it's the SMP kernel that screws up. In synchronizing shared memory access between processes most likely. And not necessarily in the exception handler.

**Comment by dmik** [comment]
I've also tried the 14.106 kernel under my older eCS install — the results are the same.

**Comment by StevenLevine** [comment]
I'm pretty sure we have discussed this before.  This is not the kernel screwing up.  There is no serialization between exception handlers running on separate cores  If you need serialization in your exception handler, it's a DIY project.  This is why exceptq has hack-ish code to detect if it  is running on some other core and avoid problems.

That said, it appears you do have serialization, just not quite the required serialization.

Is it possible that the failing thread is given access to the page before the page before the content is completely read in from disk?

**Comment by dmik** [comment]
Steven, I don't remember discussing it under the same angle, but anyway. Yes, despite the fact that I use the global process-wide mutex, there is one hole where racing can occur. As you correctly notice in your email,  there is `DosSetMem(... PAG_COMMIT)` that happens at step 4 from the comment above, i.e. inside the exception handler which is entered to read the new page from the file on its first access. We use uncommitted pages specifically for that purpose — to have the exception handler triggered, that will then commit the page and populate its contents from the file.

Both `DosSetMem` and `DosRead` (and all other related logic) are done while hloding the global mutex. So, If both processes happen to trigger the exception handler at the same time — which is the case when both processes try reading an uncommitted page at the same time and both enter the exception handler before the first `DosSetMem` — all works as designed. However, if it happens so that the second process reads the page in question just a few milliseconds later, right **after** the first process called `DosSetMem` on it but **before** it called `DosRead` to fill it with real data, a trouble occurs. The second process will continue with reading bytes from the page which will effectively return either zeroes or zeroes plus partial page contents if the first process is lucky enough to kick in with `DosRead` again in between. And this must be the reason of moc failures.

Given that, the only way to solve this race is to have an "atomic" operation that will perform `DosSetMem` and `DosRead` in one go, preventing any other process from reading this page's bytes in between. From what I see, the only way to achieve it from LIBCx is to completely stop execution of all other cores for the duration of this `DosSetMem` and `DosRead` pair. So, my question is: is such a thing possible in the OS/2 kernel at all? If not, can it be made possible via a patch or a driver?

If the answer to both questions is "no", then we are in trouble and will have to document concurrent multiprocess usage of MAP_SHARED regions as unsupported behavior.

**Comment by StevenLevine** [comment]
You might be able to use the ACPI daemon to stop the other cores, but the chance of deadlocking the system would be rather large.  Also, the race is not gone on single core systems.  The probability of it occurring is just very low.

The solution I would investigate is removing the other processes access permissions to the shared object before adjusting the access permissions to the shared object in the exception handler.  If you can do this, the other processes will be forced into the exception handler if an access attempt occurs and all will be well.

**Comment by dmik** [comment]
Okay, you are perhaps right about the single core (unless it is somehow guaranteed in the kernel that it doesn't do context switches while in exception handlers or such). Then I need a way not just to disable cores but to disable any context switching at all. Globally. Something like `DosEnterCritSec` but system-wide.

Regarding your proposed solution. It isn't practically possible because reading from the memory block returned by `mmap` may happen at any time and there is no way for LIBCx to kick in other than the exception handler. What I can try to do is send a signal to other LIBCx processes with e.g. `DosKillProcess` and then make them freeze from their exception handlers until the initiating process finishes with `DosSetMem` and `DosRead`. I need to give it a try.

**Comment by StevenLevine** [comment]
There is no way to disable all context switching in the way you are asking.  If you think about it you will realize that without context switching it  is unlikely that you could even complete the DosRead.

I don't think you quite understand my idea.  The problem is not reading from the memory block.  The problem is reading from a block and not being forced into the exception handler.  Let's say we call the exception handler that needs to do the DosRead HandlerA.    If before HandlerA sets PAG_WRITE , it loops though the other processes that are sharing the same page and issues a DosGiveSharedMem removing PAG_READ and PAG_WRITE access for these processes.  Then if a thread for one of these processes attempts read or write the page, it will enter its exception handler which is what we want to happen.  Then it will wait until HandlerA finishes the DosRead, undoes the changes to the access permissions and releases the lock.

Assuming that DosGiveSharedMem suppports access permission changes that way I think it does, it will be a relatively minor change to the existing exception handler to avoid the race condition.

It's possible your technique could work, but I think will be messy to implement.  Recall XCPT_SIGNAL_KILLPROC only goes to thread 1.

**Comment by StevenLevine** [comment]
Reviewing the docs reminds me that DosGiveSharedMem must have one of PAG_READ, PAG_WRITE or PAGE_EXECUTE set, which complicates things somewhat.  However, it does allow PAG_GUARD to be set which will ensure that the exception handler is entered.

**Comment by dmik** [comment]
Regarding your idea. As you already mentioned, PAG_READ is always implied as long as the page is committed so XCPT_ACCESS_VIOLATION is not a way to go here. What about PAG_GUARD, I was considering it for a similar purpose back then and did some tests back then. My tests showed that PAG_GUARD was ignoderd for shared memory. So I put it off the table back then.

Although, I tried it again now and this time it seems to work which is kinda strange. I clearly remember it was ignored. More over, PAG_GUARD seems to be a per-process flag, similar to PAG_READ and PAG_WRITE. I tried this approach and it seemed to work very well at first (in the debug build). Applying it to the release build still makes the moc test case fail after some iterations. I need to think more, may be there is some other hole.

**Comment by dmik** [comment]
Okay, I sorted it out — one case was not covered (the original process that inially allocated the shared region). Once I fixed that my test succeeded and seems to work really well. I tried 50 (!) iterations of this heavy moc test case (see https://github.com/bitwiseworks/qtbase-os2/issues/72) and all of them worked. So I guess we sorted this particular problem.

BTW, I use PAG_GUARD in a different manner, not the one you suggested. I set this flag for the whole region of pages in all involved processes right from the mmap call that first allocates the region (or gains access to it if it's a different process). This has a significant benefit that there is no need to track processes using any given mmap. It also makes the necessary code changes really simple. And I'm also not sure that `DosGiveSharedMem` can be called more than once for any given process and/or for individual pages at all.

Also, the described race condition existed not only between different processes, but between threads of the same process as well (and, consequently, not only for shared mappings but also for private ones). The same PAG_GUARD technique covers these races too. I didn't even have to use any additional logic for that, once setting PAG_GUARD for the whole region was in place.

There is a price for this approach though. Now, the frist access to each page of a file-backed mapping in any process will cause a XCPT_GUARD_PAGE_VIOLATION exception to happen, even if this page is already commited and read from disk. This also includes the process and the thread that do committing and reading (in the latter case XCPT_GUARD_PAGE_VIOLATION will happen inside `DosRead` and is vital for protecting from inter-thread races). This price looks rather small to me, however. The handler logic is quite cheap in this case — just lock the global mutex, locate the mapping to validate the address, then unlock the mutex and return-retry. Compared to reading page contents from the file it's nothing. And it's obvious that using per-page `DosGiveSharedMem` calls for all involved processes (if we assume that it works that way) would be much slower.

Note that with this fix, it makes sense to use MAP_SHARED when reading files even if inter-process communication per se is not really necessary (i.e. read files are never changed). While it would also work correctly with MAP_PRIVATE (as my moc tests showed), MAP_SHARED has one significant benefit: if there are other running LIBCx processes reading the same files with MAP_SHARED, these processes will **not** have to allocate new memory regions for their copies of the file (as they will all reuse the same shared memory region) and **not** have to read file contents again (as the first acessing process will have already done this). I.e. this will save both time and memory. In case of a large amount of files this might be a very significant benefit. Moc is a great example here as it reads a lot of headers and in case of parallel builds there are many moc processes running in parallel and opening the same headers over and over.

**Comment by dmik** [comment]
Now I have to bruch up the changes, limit PAG_GUARD to only file-backed mmaps, run the test suite, etc. and then commit the code changes.

PS. Regarding non-working PAG_GUARD. It might be that I tried it on private aliases for shared memory regions and it was there where it didn't have any effect. Anyway, I'm really glad it worked on shared memory per se as it turned out to be the simplest solution.

**Comment by dmik** [comment]
Well, my testing shows that this solution does not work well for inter-thread races within the single process if there are more than 2 concurring threads. The reason is simple: PAG_GUARD is a per process so the first thread that gets out of order (i.e. between `DosSetMem(PAG_COMMIT)` and `DosRead`) resets this flag and if there are other threads doing the same, they are free to go (i.e. read corrupted data). I guess that `DosEnterCritSec` will help to protect from this.

**Comment by dmik** [comment]
Well, using `DosEnterCritSec` gives a very strange effect: `DosRead` of the first page takes 30 seconds making all other threads get stuck for that long of course. The OS/2 CPREF is very vague about OS/2 APIs with which it's safe to use `DosEnterCritSec`. Perhaps, `DosRead` is not one of them.

**Comment by dmik** [comment]
This is somehow connected to the number of concurrent threads. The slowdown is only seen if there are 100 threads or such. If I limit it to 5 threads, then it's not noticeable. Perhaps, the OS/2 task scheduler is not optimized to stopping 100 threads at once or such. Just a guess, of coures. Other than this slowdown, all works correctly with `DosEnterCritSect`.

**Comment by dmik** [comment]
Well, tests of the release build show that it works w/o any delays. Debug delays can be also contributed by intensive logging. I will go on with this approach for now.

**Comment by dmik** [comment]
Writing a decent test case for both multi-thread and mupti-process races took some time. But now it's done and all works really well. Steven, you are welcome to review.

**Comment by StevenLevine** [comment]
Miscellaneous comments follow...

I pretty sure you can call DosGiveSharedMem more than once for a process.  However, I have not testcased this.

All the shared memory API are per process.  This does require that the code deal with other threads in the current process, but  DosEnterCritSec and the global lock will take care of this, as you have found.

The docs claim PAG_GUARD applies only to what the docs calls a  "private memory object."  This looks like a doc error to me, based on what is working for you.

Debug logging is always going to skew performance.  Printf debugging is notoriously intrusive.

FWIW, exiting from a critical section is similar to posting an event semphore in that multiple threads  can go into the ready state at the same time a this can cause the scheduler extra work while rebuilding the dispatch queues.  On an SMP system, this work might trigger IPIs.  That said, the kernel code given its age is probably better than we have a right to expect.

One thing I do wonder about sometimes is how our mmap performance compares to other platforms.

**Comment by dmik** [comment]
Okay, thanks for your comments.

Regarding performance. I expect it to be slower because we do all the dirty work in R3 w/o access to kernel sync primitives so that we have to use high-level mutexes and semaphores (which is especially expensive in case of named shared ones) etc. which definitely generates some overhead. Overall results should be, however, comparable (i.e. less than two times slower) as in general we use a similar approach to implement this (I believe so based on theory at least — I have not studied the Linux impl closely).
